### PR TITLE
Sprint 1

### PR DESCRIPTION
**Pull request by FylSanya**
### **What i've done (like one popular band sang (rip in peace)):**

1. Add different apps for Items, Lots, Auctions and User
2. Create models for them
3. Wrote serializers with pagination, filters, validation, blackjack and /*bad word*/
4. Declare some routes (many on them are useless, but ~~who cares~~  they are for debug convenience)
5. Add such good thing like Docker (~~i hate it~~), poetry, PostgresSQL; connect them together.

It is my first pull request, don't judge me strictly.

### CODE REVIEW

File: auction/urls.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from django.urls import path
+from rest_framework.urlpatterns import format_suffix_patterns
+from . import views
+
+urlpatterns = [
+    path('', views.AuctionList.as_view()),
+    path('<int:pk>/', views.AuctionDetail.as_view()),
+]
+
+urlpatterns = format_suffix_patterns(urlpatterns)
```
Comment (greatleroy) [inline auction/urls.py:10]: What is the main idea for using `format_suffix_patterns` ?

File: user/serializers.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    class Meta:
+        model = User
+        fields = [
+            'id',
+            'username',
+            'password'
+        ]
+
+
+class CreateUserSerializer(serializers.ModelSerializer):
```
Comment (greatleroy) [inline user/serializers.py:15]: DRY: `UserSerializer` and `CreateUserSerializer`

File: auction/views.py [hunk 3]
```diff
@@ -1,3 +1,19 @@
 from django.shortcuts import render
+from rest_framework import generics, filters
+from . import serializers
+from .models import Auction
 
-# Create your views here.
+
+class AuctionList(generics.ListAPIView):
```
Comment (greatleroy) [inline auction/views.py:7]: ...thinking about `viewsets` instead of `generics`

File: auction/serializers.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            'type',
+            'reserve_price',
+            'buy_now_price',
+            'frequency',
+            'duration'
+        ]
+
+    def to_representation(self, instance):
+        data = super().to_representation(instance)
+        if not data['frequency']:
```
Comment (greatleroy) [inline auction/serializers.py:21]: As I said -- removing an empty item is a bad idea.

File: auction/urls.py (hunk 1 above)
Comment (FylSanya) [reply auction/urls.py:10]: In my case this string of code are useless. I'll delete it.

File: user/serializers.py (hunk 2 above)
Comment (FylSanya) [reply user/serializers.py:15]: Forgot to delete it.

File: auction/serializers.py (hunk 4 above)
Comment (FylSanya) [reply auction/serializers.py:21]: Understood why, i'll fix that.

File: auction/views.py [hunk 5]
```diff
@@ -1,3 +1,19 @@
...
+from rest_framework.response import Response
+from . import serializers
+from .models import Auction
 
-# Create your views here.
+
+class AuctionViewSet(viewsets.ViewSet):
+
+    def list(self, request):
+        queryset = Auction.objects.all()
```
Comment (FylSanya) [inline auction/views.py:11]: add pagination

File: internet_auction/settings.py [hunk 6]
```diff
@@ -120,16 +107,70 @@
...
+CELERY_RESULT_BACKEND = "redis://redis:6379"
+CELERY_ACCEPT_CONTENT = ["application/json"]
+CELERY_TASK_SERIALIZER = "json"
+CELERY_RESULT_SERIALIZER = "json"
+CELERY_TIMEZONE = "Europe/Moscow"
+
+EMAIL_HOST = 'smtp.gmail.com'
+EMAIL_PORT = 587
+EMAIL_HOST_USER = 'incredible.auction.django@gmail.com'
+EMAIL_HOST_PASSWORD = 'aucDjango'
```
Comment (FylSanya) [inline internet_auction/settings.py:143]: move auth data to env file

File: internet_auction/settings.py [hunk 7]
```diff
@@ -120,16 +107,70 @@
...
+    'USER_ID_CLAIM': 'user_id',
+    'USER_AUTHENTICATION_RULE': 'rest_framework_simplejwt.authentication.default_user_authentication_rule',
+
+    'AUTH_TOKEN_CLASSES': ('rest_framework_simplejwt.tokens.AccessToken',),
+    'TOKEN_TYPE_CLAIM': 'token_type',
+
+    'JTI_CLAIM': 'jti',
+
+    'SLIDING_TOKEN_REFRESH_EXP_CLAIM': 'refresh_exp',
+    'SLIDING_TOKEN_LIFETIME': timedelta(minutes=5),
```
Comment (FylSanya) [inline internet_auction/settings.py:171]: check what is it

File: internet_auction/settings.py [hunk 8]
```diff
@@ -120,16 +107,70 @@
...
+    'ALGORITHM': 'HS256',
+    'SIGNING_KEY': SECRET_KEY,
+    'VERIFYING_KEY': None,
+    'AUDIENCE': None,
+    'ISSUER': None,
+
+    'AUTH_HEADER_TYPES': ('JWT',),
+    'AUTH_HEADER_NAME': 'HTTP_AUTHORIZATION',
+    'USER_ID_FIELD': 'id',
+    'USER_ID_CLAIM': 'user_id',
```
Comment (FylSanya) [inline internet_auction/settings.py:162]: check user_id field

File: lot/models.py [hunk 9]
```diff
@@ -1,22 +1,35 @@
...
-    item = models.OneToOneField(
+    item = models.ForeignKey(
         Item,
         on_delete=models.CASCADE,
     )
-    auction_type = None
+    auction = models.OneToOneField(
+        Auction,
+        on_delete=models.CASCADE,
+        null=False,
```
Comment (FylSanya) [inline lot/models.py:16]: check difference between null and blank

File: lot/serializers.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    class Meta:
+        model = Lot
+        fields = [
+            'id',
+            'item',
+            'auction',
+            'owner'
+        ]
+
+    def create(self, validated_data):
```
Comment (FylSanya) [inline lot/serializers.py:24]: add transaction atomic

File: lot/tasks.py [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import os
+from django.core.mail import send_mail
+from internet_auction.settings import EMAIL_HOST_USER
+
+
+def send_mail_to_user(message, destination):
```
Comment (FylSanya) [inline lot/tasks.py:6]: Create function in separate file (email_utils)
use tasks.py as tasks for celery function.

File: lot/views.py [hunk 12]
```diff
@@ -1,3 +1,38 @@
...
+        serializer = self.get_serializer(lot)
+        return Response(serializer.data)
+
+    def create(self, request):
+        serializer = self.get_serializer(
+            data=request.data,
+            context={'request': request}
+        )
+        serializer.is_valid(raise_exception=True)
+        # send_mail_to_user(json.dumps(request.data), '0bygagashka0@gmail.com')
```
Comment (FylSanya) [inline lot/views.py:36]: use `transaction.on_commit` for send_email_to_user

File: user/serializers.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class UserSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = User
+        fields = ('username',)
+
+
+class UserSerializerWithToken(serializers.ModelSerializer):
```
Comment (FylSanya) [inline user/serializers.py:13]: `serializers.Serializer` instead of `ModelSerializer`

**Pull request merged**